### ISSUE TITLE
Question about learning rate and epochs

### ISSUE THREAD

**Issue by luffycodes** [jondurbin/airoboros#19]
Hello, I was wondering if you could please let me know some of the hyperparameters used? e.g. learning rate, batch size, epochs? Do you choose the epochs such the final training loss is 0?

[here]{https://huggingface.co/jondurbin/airoboros-13b-gpt4-1.4/blob/main/README.md?code=true#L7} you mention you re-uploaded with a slightly earlier checkpoint, which seems perhaps a little less overfit. May I know how you measure the overfit? Do you split the dataset into training and validation?

**Comment by jondurbin** [comment]
By default, FastChat splits the dataset into 0.98 train and 0.02 validation, which is almost too small to be useful in many ways.  Increasing the validation split is potentially risky however, since there are likely many really interesting gems in the datasets that could be randomly labeled for test and not used in training.

I previously would look at the validation loss, but now honestly it's just trial and error testing various checkpoints and how well they generalize.  I also test various training methods, learning rates, epochs, etc., e.g.:
https://huggingface.co/jondurbin/airoboros-l2-13b-gpt4-m2.0/discussions/1#64d000fc59503263d9f00845

You can find the scripts I used for various model sizes here:
https://gist.github.com/jondurbin

I've found that for 13b models, this works fairly well:
```
export BASE_MODEL=llama-2-13b-hf
export WANDB_PROJECT=airoboros-13b-gpt4-2.0
export PYTHONPATH=FastChat

deepspeed --num_nodes=1 --num_gpus=4 FastChat/fastchat/train/train_mem.py \
  --model_name_or_path $BASE_MODEL \
  --data_path as_conversations.json \
  --output_dir $WANDB_PROJECT \
  --num_train_epochs 2.5 \
  --evaluation_strategy "no" \
  --save_steps 100 \
  --save_total_limit 1 \
  --learning_rate 1e-5 \
  --weight_decay 0. \
  --warmup_ratio 0.04 \
  --lr_scheduler_type "cosine" \
  --logging_steps 1 \
  --bf16 \
  --per_device_train_batch_size 12 \
  --deepspeed deepspeed.json \
  --model_max_length 4096 \
  --gradient_checkpointing True \
  --lazy_preprocess True
  ```
  
  deepspeed.json:
  ```
  {
  "gradient_accumulation_steps": "auto",
  "gradient_clipping": "auto",
  "train_batch_size": "auto",
  "train_micro_batch_size_per_gpu": "auto",
  "bf16": {
    "enabled": true
  },
  "zero_optimization": {
    "stage": 2,
    "offload_optimizer": {
      "device": "cpu"
    },
    "contiguous_gradients": true,
    "overlap_comm": true,
    "reduce_scatter": true,
    "reduce_bucket_size": 5e8,
    "allgather_bucket_size": 5e8
  }
}
```
  
For that instance, the training loss ended at around 0.23.  I've had decent results with as few as 2 epochs, and probably less than that would be fine, depending on your dataset size.

**Comment by luffycodes** [comment]
Thanks so much for such valuable information ! :)  This is very helpful..
Also, just one more question: do I need to half the per_device_train_batch_size to 6 if i have num_nodes=2 or does per_device_train_batch_size remain same?

**Comment by jondurbin** [comment]
I would start there, then bump it up a bit if vram usage is below 90-95%.

**Comment by luffycodes** [comment]
Okay thank you !
